pkg/machinery/client/grpc_connection_wrapper.go

```diff
@@ -6,33 +6,108 @@ package client
 
 import (
 	"context"
+	"runtime"
 
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
 	"google.golang.org/grpc/metadata"
 )
 
 type grpcConnectionWrapper struct {
-	*grpc.ClientConn
+	conn *grpc.ClientConn
 
 	clusterName string
 }
 
 func newGRPCConnectionWrapper(clusterName string, conn *grpc.ClientConn) *grpcConnectionWrapper {
-	return &grpcConnectionWrapper{
-		ClientConn:  conn,
+	res := &grpcConnectionWrapper{
+		conn:        conn,
 		clusterName: clusterName,
 	}
+
+	runtime.SetFinalizer(res, func(c *grpcConnectionWrapper) { c.Close() }) //nolint:errcheck
+
+	return res
+}
+
+func (c *grpcConnectionWrapper) Close() error {
+	err := c.conn.Close()
+
+	runtime.SetFinalizer(c, nil)
+	runtime.KeepAlive(c)
+
+	return err
+}
+
+// NewStream begins a streaming RPC.
+func (c *grpcConnectionWrapper) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	res, err := c.conn.NewStream(c.appendMetadata(ctx), desc, method, opts...)
+
+	runtime.KeepAlive(c)
```

Comment:
this starts to feel more and more hacky solution to me, and more fragile.

Also I don't see how it fixes anything, as in this function the result (`grpc.ClientStream`) can run for as long as possible once this function returns, and still the problem is there.

Why can't we spend the effort other way around and find where the connection is not closed? It gives us immediate benefit? Easy ways there, like adding an assertion using `goleak` that no goroutines are left in each unit-test, and just simply inspecting the code. 

Even if we find a finalizer solution which works (which I doubt), it feels that fixing the core problem is more beneficial immediately.

Reply:
Yeah, you right. I think we should try something different.